agent/xds/listeners_ingress.go

```diff
@@ -154,21 +137,101 @@ func (s *ResourceGenerator) makeIngressGatewayListeners(address string, cfgSnap
 	return resources, nil
 }
 
-func resolveListenerSDSConfig(cfgSnap *proxycfg.ConfigSnapshot, listenerCfg structs.IngressListener) (*structs.GatewayTLSSDSConfig, error) {
+func makeDownstreamTLSContextFromSnapshotListenerConfig(cfgSnap *proxycfg.ConfigSnapshot, listenerCfg structs.IngressListener) (*envoy_tls_v3.DownstreamTlsContext, error) {
+	var downstreamContext *envoy_tls_v3.DownstreamTlsContext
+
+	tlsContext, err := makeCommonTLSContextFromSnapshotListenerConfig(cfgSnap, listenerCfg)
+	if err != nil {
+		return nil, err
+	}
+
+	if tlsContext != nil {
+		downstreamContext = &envoy_tls_v3.DownstreamTlsContext{
+			CommonTlsContext:         tlsContext,
+			RequireClientCertificate: &wrappers.BoolValue{Value: false},
+		}
+	}
+
+	return downstreamContext, nil
+}
```

Comment:
In this code, if `tlsContext` is ever nil, the function will return `nil, nil` which is a pattern to avoid.

Idiomatically, `tlsContext` should never be nil if err is nil (unless documented clearly otherwise) so I personally don't think there's a need for defensive coding here.

If you _do_ want to guard against tlsContext being nil, we should throw an error like
```go
if tlsContext == nil {
        return nil, errors.New("cannot make tlsContext")
}
```

so that all codepaths return either `nil, err` or `something, nil`

Reply:
🤔 I'm looking a little more closely at the code called by this function, and it does appear it could be possible for `makeCommonTLSContextFromSnapshotListenerConfig` to return a `nil` `CommonTLSContext` in the case where SDS is not enabled and `gatewayTLSCfg.Enabled` is false or unset.

I _think_ this is a valid case where it simply won't configure a TLS `transportSocket` for the listener in the generated Envoy config (as seen in the port 8080 listener in `agent/xds/testdata/listeners/ingress-with-single-tls-listener.envoy-1-20-x.golden`), not an error case.